src/sv_simulator.jl

```diff
@@ -318,3 +326,227 @@ function partial_trace(
     return final_ρ
 end
 partial_trace(sim::StateVectorSimulator, output_qubits = collect(0:qubit_count(sim)-1)) = partial_trace(state_vector(sim), output_qubits)
+
+##################################
+# Measurement Related Operations #
+##################################
+
+"""
+    get_measurement_probabilities(state::AbstractVector{<:Complex}, qubit::Int)
+
+Calculate measurement probabilities for a qubit in a state vector.
+Returns [P(0), P(1)] for the given qubit.
+"""
+function get_measurement_probabilities(state::AbstractVector{<:Complex}, qubit::Int)
+    n_qubits = Int(log2(length(state)))
+    
+    # Check if qubit index is valid
+    if qubit < 0 || qubit >= n_qubits
+        error("Invalid qubit index: $qubit. Must be between 0 and $(n_qubits-1)")
+    end
+    
+    endian_qubit = n_qubits - qubit - 1
+    
+    # Calculate probabilities for 0 and 1 outcomes
+    prob_0 = 0.0
+    prob_1 = 0.0
+    
+    # Iterate through all amplitudes
+    for i in 0:(length(state)-1)
+        # Check if the qubit is 0 or 1 in this basis state
+        qubit_val = (i >> endian_qubit) & 1
+        prob = abs2(state[i+1])  # Probability of this basis state
+        
+        if qubit_val == 0
+            prob_0 += prob
+        else
+            prob_1 += prob
+        end
+    end
+    
+    return [prob_0, prob_1]
+end
+
+"""
+    _apply_reset(state::AbstractVector{<:Complex}, target::Int)
+
+Reset a qubit to |0⟩ state in a state vector.
+This projects the state to the |0⟩ state of the target qubit and then normalizes.
+Returns the new state.
+"""
+function _apply_reset(state::AbstractVector{<:Complex}, target::Int)
+    # First measure the qubit (project to either |0⟩ or |1⟩)
+    probs = get_measurement_probabilities(state, target)
+    outcome = rand() < probs[1] ? 0 : 1
+    apply_projection!(state, target, outcome)
+    
+    # If outcome was |1⟩, apply X gate to make it |0⟩
+    if outcome == 1
+        n_qubits = Int(log2(length(state)))
+        endian_target = n_qubits - target - 1
+        
+        # Apply X gate to flip the target qubit
+        for i in 0:(length(state)÷2-1)
+            lower_ix = pad_bit(i, endian_target) + 1
+            higher_ix = flip_bit(lower_ix - 1, endian_target) + 1
+            
+            # Swap amplitudes
+            temp = state[lower_ix]
+            state[lower_ix] = state[higher_ix]
+            state[higher_ix] = temp
```

Comment:
Is it possible to reuse the X-gate logic here? Alternatively, change the existing X-gate logic to do this.